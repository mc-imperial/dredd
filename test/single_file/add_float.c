int main() {
  float x = 5.235;
  float y = 754.34623;
  float z;
  z = x + y;
  return 0;
}
