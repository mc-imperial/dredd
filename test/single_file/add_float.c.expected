#include <inttypes.h>
#include <stdlib.h>
#include <string.h>
static int __dredd_some_mutation_enabled = 1;
static int __dredd_enabled_mutation(int local_mutation_id) {
  static int initialized = 0;
  static uint64_t enabled_bitset[1];
  if (!initialized) {
    int some_mutation_enabled = 0;
    const char* dredd_environment_variable = getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable) {
      char* temp = malloc(strlen(dredd_environment_variable) + 1);
      strcpy(temp, dredd_environment_variable);
      char* token;
      token = strtok(temp, ",");
      while(token) {
        int value = atoi(token);
        int local_value = value - 0;
        if (local_value >= 0 && local_value < 33) {
          enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
          some_mutation_enabled = 1;
        }
        token = strtok(NULL, ",");
      }
      free(temp);
    }
    initialized = 1;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64));
}

static int __dredd_replace_expr_int_uoi_optimised_zero(int arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ~(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return 1;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -1;
  return arg;
}

static float __dredd_replace_expr_float_lvalue(float* arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return (*arg);
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return ++((*arg));
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return --((*arg));
  return (*arg);
}

static float __dredd_replace_expr_float(float arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return 0.0;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return 1.0;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -1.0;
  return arg;
}

static float __dredd_replace_binary_operator_Assign_float_float(float* arg1, float arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return (*arg1) = arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return (*arg1) += arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return (*arg1) /= arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return (*arg1) *= arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return (*arg1) -= arg2;
  return (*arg1) = arg2;
}

static float __dredd_replace_binary_operator_Add_float_float(float arg1, float arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1 + arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1 / arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1 * arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1 - arg2;
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg2;
  return arg1 + arg2;
}

int main() {
  float x = __dredd_replace_expr_float(5.235, 0);
  float y = __dredd_replace_expr_float(754.34623, 3);
  float z;
  if (!__dredd_enabled_mutation(28)) { __dredd_replace_binary_operator_Assign_float_float(&(z) , __dredd_replace_expr_float(__dredd_replace_binary_operator_Add_float_float(__dredd_replace_expr_float(__dredd_replace_expr_float_lvalue(&(x), 6), 8) , __dredd_replace_expr_float(__dredd_replace_expr_float_lvalue(&(y), 11), 13), 16), 21), 24); }
  if (!__dredd_enabled_mutation(32)) { return __dredd_replace_expr_int_uoi_optimised_zero(0, 29); }
}
